### PR TITLE
fix: apply serializer to `Error` instance for thrown snapshot

### PR DESCRIPTION
**Pull request by hi-ogawa**
### Description

Closes https://github.com/vitest-dev/vitest/issues/3655

As explained in https://github.com/vitest-dev/vitest/issues/3655#issuecomment-1784471527, currently vitest hard-codes specific behavior to extract `Error.message` for `Error` instance when it's snapshot-ed via `toThrowErrorMatching...`, which prevents error from being applied with user-provided serializer.
This PR removes this special behavior, so that `Error` instance's snapshot is a consistent regardless of whether its snapshote-ed via `toMatchInlineSnapshot` or `toThrowErrorMatchingInlineSnapshot`.

Due to the removal of `Error.message` extraction, the default snapshot format changes in a following way

```tsx
// before
  expect(() => {
    throw new Error('omega')
  }).toThrowErrorMatchingInlineSnapshot(`"ometa"`)

// after
  expect(() => {
    throw new Error('omega')
  }).toThrowErrorMatchingInlineSnapshot(`[Error: omega]`)
```

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by sheremet-va** [comment]
This looks fine to me. We already break _evey_ snapshot with default values in Vitets 1.0, so might as well.

**Review by sheremet-va — APPROVED** [review]

**Comment by hi-ogawa** [comment]
Hi, thanks for confirming. Good to know that the breaking change is acceptable now.
I'll continue with this PR to cleanup some code and also adjust some test cases.

**Review by sheremet-va — APPROVED** [review]

**Comment by sheremet-va** [comment]
Would also be nice to add this to the migration from Jest guide. Jest uses `error.message` currently.

**Comment by hi-ogawa** [comment]
> Would also be nice to add this to the migration from Jest guide. Jest uses `error.message` currently.

Good point. I didn't think about jest compatibility. I'll take a look at the documentation.

**Comment by hi-ogawa** [comment]
I updated the documentation to reflect the change in this PR.
I wasn't sure which, but instead of `guide/migration.md`, I updated `guide/snapshot.md` since it has a dedicated "Difference from Jest" section.

File: docs/guide/snapshot.md [hunk 1]
```diff
@@ -237,5 +237,32 @@ exports[`toThrowErrorMatchingSnapshot: hint 1`] = `"error"`;
...
+  //
+  // in Jest
+  //
+
+  expect(new Error('error')).toMatchInlineSnapshot(`[Error: error]`)
+
+  // Jest snapshots `Error.message` for `Error` instance
+  expect(() => {
+    throw new Error('error')
+  }).toThrowErrorMatchingInlineSnapshot(`"error"`)
```
Comment (hi-ogawa) [inline docs/guide/snapshot.md:256]: I verified this behavior in https://stackblitz.com/edit/jest-example-t1qnq1?file=sum.test.js

<details><summary>reveal screenshot</summary>

image


</details>

**Pull request merged**